### PR TITLE
snap/quota: add values for journal quotas (journal quota 2/n)

### PR DESCRIPTION
**Pull request by Meulengracht**
The second PR in the journal quota support series; add the neccessary support for journal quotas in snap/quota/quota.go

### CODE REVIEW

File: snap/quota/quota.go [hunk 1]
```diff
@@ -57,6 +57,24 @@ type GroupQuotaCPU struct {
 	AllowedCPUs []int `json:"allowed-cpus,omitempty"`
 }
 
+// GroupQuotaJournal contains the supported limits for journald. Any limit set here
+// applies only the quota group itself. Journal limits will not be inherited by the
```
Comment (mardy) [inline snap/quota/quota.go:61]: nitpick: ...only *to* the...

**Review by mardy — APPROVED** [review]
LGTM, thanks!

File: snap/quota/quota.go [hunk 2]
```diff
@@ -57,6 +57,24 @@ type GroupQuotaCPU struct {
...
+	// Size is the maximum allowed size of the journal for the group.
+	// If the size is set below current usage, systemd will automatically treat
+	// the current usage of the journald namespace as the minimum limit and
+	// render whatever set here ineffective. The maximum allowed size for
+	// journald namespaces is 4GB. A value of 0 here means no limit is present.
+	Size quantity.Size `json:"size,omitempty"`
+
+	// RateCount/RatePeriod determines the maximum rate of journal writes for
+	// the group. The count is the number of journal messages that can be written
+	// in each period. 0 Values here means there is no limit currently set.
```
[the remark this answers is not in the document]
Comment (Meulengracht) [reply snap/quota/quota.go:73]: You caught an oversight here, thanks for that! I'll update this to reflect that the resolution is microseconds. See PR https://github.com/snapcore/snapd/pull/11738

File: snap/quota/quota.go (hunk 1 above)
Comment (Meulengracht) [reply snap/quota/quota.go:61]: Ah, thanks!

File: snap/quota/quota.go [hunk 3]
```diff
@@ -142,6 +165,15 @@ func (grp *Group) GetQuotaResources() Resources {
...
 	}
+	if grp.JournalLimit != nil {
+		resourcesBuilder.WithJournalNamespace()
+		if grp.JournalLimit.Size != 0 {
+			resourcesBuilder.WithJournalSize(grp.JournalLimit.Size)
+		}
+		if grp.JournalLimit.RateCount != 0 && grp.JournalLimit.RatePeriod != 0 {
+			resourcesBuilder.WithJournalRate(grp.JournalLimit.RateCount, grp.JournalLimit.RatePeriod)
+		}
+	}
```
Comment (stolowski) [inline snap/quota/quota.go:176]: Can w test it, or do you have tests in another PR?

File: snap/quota/quota.go [hunk 4]
```diff
@@ -672,6 +712,18 @@ func (grp *Group) UpdateQuotaLimits(resourceLimits Resources) error {
...
+	if resourceLimits.Journal != nil {
+		if grp.JournalLimit == nil {
+			grp.JournalLimit = &GroupQuotaJournal{}
+		}
+		if resourceLimits.Journal.Size != nil {
+			grp.JournalLimit.Size = resourceLimits.Journal.Size.Limit
+		}
+		if resourceLimits.Journal.Rate != nil {
+			grp.JournalLimit.RateCount = resourceLimits.Journal.Rate.Count
+			grp.JournalLimit.RatePeriod = resourceLimits.Journal.Rate.Period
```
Comment (stolowski) [inline snap/quota/quota.go:724]: Same question.

**Review by stolowski — COMMENTED** [review]
Looks good, question about tests.

File: snap/quota/quota.go (hunk 3 above)
Comment (Meulengracht) [reply snap/quota/quota.go:176]: I have added a unit test for this now, thanks for catching this!

File: snap/quota/quota.go (hunk 4 above)
Comment (Meulengracht) [reply snap/quota/quota.go:724]: Same, fixed!

File: snap/quota/quota.go (hunk 1 above)
Comment (Meulengracht) [reply snap/quota/quota.go]: Fixed

File: snap/quota/quota.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Meulengracht) [reply snap/quota/quota.go:74]: I have now updated this PR to correctly reflect that RatePeriod is a time.Duration, I'm not sure if any changes to the doc string is required, how do you feel @mardy ?

**Review by mvo5 — APPROVED** [review]
Thank you, this looks fine.
